File: all-submissions/1036-rotting-oranges/solution.java
class Solution {
    class State {
        int row;
        int col;
        int mins;
         State(int row, int col, int mins) {
            this.row = row;
            this.col = col;
            this.mins = mins;
        }
    }

    public int orangesRotting(int[][] grid) {
        int[][] directions = new int[][]{{0,1}, {1,0}, {0,-1}, {-1,0}};
        Queue<State> queue = new LinkedList<>();
        int m = grid.length;
        int n = grid[0].length;
        boolean[][] seen = new boolean[m][n];

        //get all rotten oranges in queue
        for(int i=0; i<m ;i++) {
            for (int j=0;j<n;j++) {
                if(grid[i][j] == 2) {
                    queue.add(new State(i,j,-1));
                    seen[i][j] = true;
                }
            }
        }
        int totalMins = 0;
        while(!queue.isEmpty()){
            State state = queue.poll();
            int row = state.row; int col = state.col; int mins= state.mins;
            for(int[] direction: directions) {
                int nextRow = row+direction[0];
                int nextCol = col+direction[1];
                if(isValid(nextRow, nextCol, grid) && !seen[nextRow][nextCol]){
                    queue.add(new State(nextRow, nextCol, mins+1));
                    grid[nextRow][nextCol] = 2;
                    seen[nextRow][nextCol] = true;
                }
            }
            totalMins = mins+1;
        }

        for(int i=0; i<m ;i++) {
            for (int j=0;j<n;j++) {
                if(grid[i][j] == 1) {
                    return -1;
                }
            }
        }

        return totalMins;
    }

    boolean isValid( int row, int col, int[][]grid) {
        return (row>=0 && row<grid.length 
        && col>=0 && col<grid[0].length 
        && grid[row][col] == 1);
    }
}

File: all-submissions/1036-rotting-oranges/solution.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        if not grid: 
            return -1
        ROWS = len(grid)
        COLS = len(grid[0])
        queue= deque()
        fresh = 0
        min_mins = 0
        directions = [(0,1),(1,0), (-1,0),(0,-1)]
        for i in range(ROWS):
            for j in range(COLS): 
                if grid[i][j] == 2:
                    queue.append((i,j))
                elif grid[i][j] == 1:
                    fresh += 1
        
        # edge case: no fresh oranges
        if fresh == 0:
            return 0
        while queue and fresh > 0:
            min_mins+=1
            queue_len = len(queue)
            for  _ in range(queue_len):
                row, col = queue.popleft()
                for dr,dc in directions:
                    nr = row+dr
                    nc = col +dc
                    if 0<=nr<ROWS and 0<=nc<COLS and grid[nr][nc] == 1:
                        fresh -=1
                        grid[nr][nc] = 2
                        queue.append((nr, nc))
            
        
        return min_mins if fresh == 0 else -1

